rfcs/20250430-host-side-scalars/README.md

```diff
@@ -0,0 +1,205 @@
+same device as the one they execute on (with the exception of reorder).
+
+A new request is to pass scalar values that reside on host device to GPU kernels
+at execution time. An example is the scale for SDPA operator in Pytorch [^1],
+which is a python variable on the host and not a `torch.tensor` object.
+
+There are two challenges here:
+- How to specify to primitive descriptor creation that a given input is a host
+    scalar?
+    - Currently, we don't have any engine specified to input memory descriptors.
+- How to pass host-side scalars to the execute function?
+    - Currently, we only take `dnnl::memory` objects.
+
+## Options for introducing host-side scalars support
+
+### **Recommended option**: Expose a new memory descriptor kind
+
+The proposal introduces host-side scalars using a dedicated memory descriptor
+and memory creation function.
+
+This **new memory descriptor** is lightweight: user only specify the scalar data
+type. A special format kind `dnnl_format_kind_host_scalar` marks it as a
+host-side scalar, so that implementations can detect this and pass the scalar
+as a kernel parameter (see SDPA POC [^2])
+(this property also accessible by the user via `dnnl_memory_desc_query` API).
+
+When creating a memory object for a host-side scalar via the **new memory creation function**,
+the value is stored internally, so user does not need to manage its lifetime.
+No engine is required in the API, so that there is no need for a CPU engine
+or building the library with CPU support.
+Note that since the memory object is host-based, GPU implementations must be updated to
+detect and handle host-side scalars by passing the value as a kernel
+argument, not by accessing it via memory. By default, all the GPU implementations
+should return an error for this memory unless explicit support is provided.
+
+```C
+// C API
+
+// Scalar value
+float value = 42.0f;
+
+// Create a memory descriptor for a host-side scalar
+dnnl_memory_desc_t scalar_md;
+dnnl_memory_desc_create_host_scalar(
+        &scalar_md, dnnl_f32);
+
+// Create a memory object for the scalar
+dnnl_memory_t scalar_mem;
+dnnl_status_t status = dnnl_memory_create_host_scalar(&scalar_mem, dnnl_f32, &value);
+
+// Use as regular memory object in execute function...
+```
+
+```C
+// C++ API
+
+// Scalar value
+float value = 42.0f;
+
+// Create a memory object for the scalar (no engine needed)
+memory scalar_mem(memory::desc::host_scalar(memory::data_type::f32), &value);
+
+// Use as regular memory object in execute function...
+```
+
+POC for API changes is available in the draft PR [^3],
+you could check it out for simple example, API tests, and some implementation
+details (although the implementation is not complete yet).
+
+Pros:
+- Users do not need to recreate primitives when scalar values change.
+- Allows passing scalars as kernel arguments, minimizing performance overhead.
+
+Cons:
+- Ties memory descriptor to a host.
```

Comment:
this one is no more relevant since engine is no more passed.

Reply:
now removed